### DISCUSSION TITLE
[rdom]  Why is every node - SPAN wrapped?

### DISCUSSION THREAD

**Discussion by bt-3000** [thi-ng/umbrella#470] [Q&A] [answered]
```
const dom = [
  'section', {},
  ['h1', {}, `Title`]
]  
```
compiles to:
```
<section><h><span>Title</span></h></section>
```

waiting for the result:
```
<section><h>Title</h></section>
````

Do I understand correctly that rdom is not compatible with hiccup semantics?

example:

```
// does not work
const dom = ['h', 'Title']
```

```
// working
const dom = ['h', {}, 'Title']
```

**Comment by postspectacular** [comment, marked as answer]
@bt-3000 First off, rdom **is** hiccup compatible, but **always** requires the "normalized" form, i.e.`[tag, attribs, ...children]`. See [`$tree()`](https://docs.thi.ng/umbrella/rdom/functions/_tree.html) and [`$compile()`](https://docs.thi.ng/umbrella/rdom/functions/_compile.html)...

The `<span>` are used as wrappers since elements can have multiple text children, each of which might need to be updated individually. I'm aware these text nodes could have been created via `document.createTextNode()` but current approach is more uniform (and there might have been some other issues whose details escape me right now...) Can all be revisited if needed...

In any case, were you just asking out of interest or do you have a particular issue with these spans?

**Reply by bt-3000** [reply]
thanks for the answer,
there were no problems with this

**Comment by bt-3000** [comment]
found an example where it doesn't fit
```
[ 'article', {},
 [  'p', {},  `is simply dummy text...` ]
]
```

rendered
```
<article><p><span>is simply dummy text...</span></p></article>
```

**Comment by postspectacular** [comment]
@bt-3000 Unless I'm missing something here in your example, this looks _exactly_ as intended (i.e. only non-element values or directly embedded reactive values will be span-wrapped)...

- `["div", {}, 42]` => `<div><span>42</span></div>`
- `["div", {}, reactive(42).map(x => x * 10)]` => `<div><span>420</span></div>`
- `["div", {}, ["p", {}, ["br",{}]]]` => `<div><p><br/></p></div>`
